### PR TITLE
Bug 1343328 - Ensure `id` is set for all PerformanceDatum

### PR DESCRIPTION
**Pull request by klahnakoski**
This is the first step to allowing `id` loop-around for the `performance_datum` table.

### CODE REVIEW

File: treeherder/perf/migrations/0031_auto_20200427_1450.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+                ('chunk_size', models.IntegerField(default=100)),
+                ('count', models.BigIntegerField()),
+            ],
+        ),
+        migrations.AlterField(
+            model_name='performancedatum',
+            name='id',
+            field=models.IntegerField(primary_key=True, serialize=False),
+        ),
+    ]
```
Comment (klahnakoski) [inline treeherder/perf/migrations/0031_auto_20200427_1450.py:26]: In theory this migration does nothing

File: treeherder/perf/models.py [hunk 2]
```diff
@@ -227,9 +227,39 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
         return len(older_ids) or max_chunk_size
 
 
+class Counter(models.Model):
+    model = models.CharField(max_length=80, blank=False, null=False, primary_key=True)
+    chunk_size = models.IntegerField(null=False, default=100)
+    count = models.BigIntegerField(null=False)
```
Comment (klahnakoski) [inline treeherder/perf/models.py:233]: We will keep track of all IDs in the database, in chunks

File: treeherder/perf/models.py [hunk 3]
```diff
@@ -227,9 +227,39 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+
+
+COUNTERS = {}
```
Comment (klahnakoski) [inline treeherder/perf/models.py:236]: We will keep track of all IDs, in fine detail, in memory

**Comment by sarah-clements** [comment]
@klahnakoski you should assign yourself to the bugzilla bug.

File: tox.ini [hunk 4]
```diff
@@ -23,7 +23,7 @@ commands =
     pre-commit run --all-files
     sh -c "SITE_URL=https://treeherder.dev TREEHERDER_DEBUG=False ./manage.py check --deploy --fail-level WARNING"
     # Exercising running tests outside of Docker
-    pytest tests/ --ignore=tests/selenium --ignore=tests/extract
+    pytest tests/ --maxfail=1 --ignore=tests/selenium --ignore=tests/extract
```
Comment (klahnakoski) [inline tox.ini:26]: Fail fast

File: treeherder/perf/migrations/0032_add_perf_id_counters.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+                ('chunk_size', models.IntegerField(default=100)),
+                ('count', models.BigIntegerField()),
+            ],
+        ),
+        migrations.AlterField(
+            model_name='performancedatum',
+            name='id',
+            field=models.AutoField(primary_key=True, serialize=False, verbose_name='ID'),
+        ),
+    ]
```
Comment (klahnakoski) [inline treeherder/perf/migrations/0032_add_perf_id_counters.py:26]: Does not change the `performance_datum` table:

```
C:\Users\kyle\code\treeherder>python manage.py sqlmigrate perf 0032
--
-- Create model Counter
--
CREATE TABLE `perf_counter` (`model` varchar(80) NOT NULL PRIMARY KEY, `chunk_size` integer NOT NULL, `count` bigint NOT NULL);
--
-- Alter field id on performancedatum
--
```

File: treeherder/perf/models.py [hunk 6]
```diff
@@ -227,9 +229,47 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+COUNTERS_LOCK = Lock()
+
+
+def next_id(model):
+    """
+    :return:  A unique integer key for given model
+    """
+    start = 1
+    model_name = model.__name__
+    with COUNTERS_LOCK:
```
Comment (klahnakoski) [inline treeherder/perf/models.py:248]: I do not know how many threads will be here at the same time.

File: treeherder/perf/models.py [hunk 7]
```diff
@@ -227,9 +229,47 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+        count, maxx = COUNTERS.get(model_name, (start, 0))
+        if count >= maxx:
+            # we require a new block of ids
+            with transaction.atomic():
```
Comment (klahnakoski) [inline treeherder/perf/models.py:252]: I am a bit concerned that a transaction in a lock will deadlock. Django makes it hard to control the commands sent to the database, nevermind the controlling the order of those commands.  I hope that updating the `Counter` will be executed, and not accumulated in a transaction to be sent at a later time.

File: treeherder/perf/models.py [hunk 8]
```diff
@@ -227,9 +229,47 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+                db_counter.save()
+
+        COUNTERS[model_name] = (count + 1, maxx)
+        return count
+
+
 class PerformanceDatum(models.Model):
     objects = PerformanceDatumManager()
-
+    id = models.AutoField(primary_key=True, serialize=False, verbose_name='ID')
```
Comment (klahnakoski) [inline treeherder/perf/models.py:272]: Make the implicit `id` explicit.  Probably not needed.

**Comment by klahnakoski** [comment]
@armenzg   you are also invited to review, but I already ask you for many.

File: treeherder/perf/models.py (hunk 7 above)
Comment (ionutgoldan) [reply treeherder/perf/models.py:252]: This feels complicated & I still don't know if Django uses a multi threaded and/or multi processing execution model. However, Django already provides support for concurrent access to models. Even at row level.

I think you should rather use [selected_for_update()](https://docs.djangoproject.com/en/3.0/ref/models/querysets/#select-for-update) instead.
[Others](https://stackoverflow.com/questions/1030270/race-conditions-in-django) have used this mechanism as well.

File: treeherder/perf/models.py [hunk 9]
```diff
@@ -227,9 +229,47 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
         return len(older_ids) or max_chunk_size
 
 
+class Counter(models.Model):
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:232]: `null` parameter defaults to `False` [in Django](https://docs.djangoproject.com/en/3.0/ref/models/fields/#null). We should avoid being explicit about this, as this is also a general project convention.

File: treeherder/perf/models.py [hunk 10]
```diff
@@ -227,9 +229,47 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+    model = models.CharField(max_length=80, blank=False, null=False, primary_key=True)
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:233]: Similar thing for the `blank` parameter.

File: treeherder/perf/models.py (hunk 7 above)
Comment (klahnakoski) [reply treeherder/perf/models.py:252]: Django is definitely run on multiple machines and multiple processes: The biggest users are the log parsers.

Comment (ionutgoldan) [reply treeherder/perf/models.py:252]: Regarding the global `COUNTERS` variable... I'm not sure that's the best approach.
It could be read/updated not only by web requests, but by cronjobs also or even by one-off Dyno commands.

Comment (klahnakoski) [reply treeherder/perf/models.py:251]: > Regarding the global COUNTERS variable... I'm not sure that's the best approach.
It could be read/updated not only by web requests, but by cronjobs also or even by one-off Dyno commands.

This is what the `COUNTERS_LOCK` is for:  It ensures that each thread in a process has exclusive accesses the COUNTERS.

> The multi thread lock won't suffice for the cross-process problem.

That is what `transaction.atomic():` is for: Ensuring the `db_counter` is unique among all the processes that access the database.

> On SO, people advice against using global shared variables, as Django has a "shared-nothing" architecture.

I completely agree with the philosophy of the "shared nothing" architecture, unfortunately Django and Treeherder are not using that philosophy:  The `AUTO INCREMENT` columns are shared global variables.  `AUTO INCREMENT` columns should not be used:  They are convenient for small apps, but do not scale.

This PR is an attempt to avoid a large database migration, and replace the `AUTO INCREMENT` feature which provides a dense, non-overlapping, id numbers that fit in `INT32`.  Since `AUTO INCREMENT` is a global variable, it makes sense the replacement will act similar.. 

This code has two levels:  One, to serve chunks of ids to the various machines/processes, and another to coordinate the serving of individual ids to each of the threads.

**Comment by klahnakoski** [comment]
@armenzg   you are also invited to review, but I already ask you for many.

**Comment by klahnakoski** [comment]
> After letting this run over the evening. It appears to be issuing duplicate values.

I see nothing wrong with code.  So, here is what could be happening:

* The code is wrong: It is issuing duplicates 
* There is a PerformanceDatum being created without being given an `id` - I have gone over the code a few time now
* Django is ignoring the `id` provided - Stepping through the code this does not appear to be the case, the INSERT statement includes `id`
* Some other process is accessing the database using old code, so it is using `AUTO INCREMENT`, and there are conflicts. - I have not eliminated this possibility:  I am not certain what other machines may be using this.

File: treeherder/perf/models.py (hunk 7 above)
Comment (klahnakoski) [reply treeherder/perf/models.py:252]: > I think you should rather use selected_for_update() instead.

Thank you for mentioning this, I will look at that.  I am not sure if `selected_for_update()` will work when I must create a new record.

```
from django.db import transaction

entries = Entry.objects.select_for_update().filter(author=request.user)
with transaction.atomic():
    for entry in entries:
        ...
```

There is something pathological about starting a transaction outside the block declaring a transaction.

**Comment by klahnakoski** [comment]
> Some other process is accessing the database using old code, so it is using AUTO INCREMENT, and there are conflicts. - I have not eliminated this possibility: I am not certain what other machines may be using this.

Confirmed?  It seems the log parsers are not all updated when deployed to prototype:  Something is using the auto increment value.

**Comment by klahnakoski** [comment]
> I see nothing wrong with code. So, here is what could be happening:

It appears there are races between threads in a process.

```
Apr 28 15:28:15 treeherder-prototype app/worker_log_parser.3: [2020-04-28 15:28:14,847] INFO [treeherder.perf.models:265] Issue block ending 1116017292 
Apr 28 15:28:15 treeherder-prototype app/worker_log_parser.3: [2020-04-28 15:28:14,847: INFO/ForkPoolWorker-3] Issue block ending 1116017292 
Apr 28 15:28:15 treeherder-prototype app/worker_log_parser.3: [2020-04-28 15:28:14,929] INFO [treeherder.perf.models:265] Issue block ending 1116017292 
Apr 28 15:28:15 treeherder-prototype app/worker_log_parser.3: [2020-04-28 15:28:14,929: INFO/ForkPoolWorker-5] Issue block ending 1116017292 
```

**Comment by klahnakoski** [comment]
It appears the transaction isolation level may also be a problem?

```
Apr 28 15:30:29 treeherder-prototype app/worker_log_parser.3: [2020-04-28 15:30:29,372] INFO [treeherder.perf.models:265] Issue block ending 1116022592 
Apr 28 15:30:29 treeherder-prototype app/worker_log_parser.4: [2020-04-28 15:30:29,382] INFO [treeherder.perf.models:265] Issue block ending 1116022592 
Apr 28 15:30:29 treeherder-prototype app/worker_log_parser.4: [2020-04-28 15:30:29,382: INFO/ForkPoolWorker-7] Issue block ending 1116022592 
Apr 28 15:30:29 treeherder-prototype app/worker_log_parser.3: [2020-04-28 15:30:29,372: INFO/ForkPoolWorker-4] Issue block ending 1116022592 
```

**Comment by klahnakoski** [comment]
Digging deep into the Django and what it sends to the database: It does not send a `START TRANSACTION`.  I will continue looking, but it seems Django does not provide sufficient transaction control.

File: treeherder/config/settings.py [hunk 11]
```diff
@@ -159,6 +159,9 @@
         # truncated upon insertion, so Django highly recommends activating a strict mode for MySQL to
         # prevent data loss (either STRICT_TRANS_TABLES or STRICT_ALL_TABLES).
         'init_command': "SET sql_mode='STRICT_TRANS_TABLES'",
+        # Ensure we read the value inside the transaction, not some other volatile value
+        # https://docs.djangoproject.com/en/2.1/ref/databases/#mysql-isolation-level
+        'isolation_level': 'serializable',
```
Comment (klahnakoski) [inline treeherder/config/settings.py:164]: IMPORTANT: This will ensure the data read from the database will not be changed by another transaction.  It may slowdown all transactions

File: treeherder/perf/models.py (hunk 6 above)
Comment (klahnakoski) [reply treeherder/perf/models.py]: There are thread pools in the log parsers.  This lock is required

File: treeherder/perf/models.py (hunk 7 above)
Comment (ionutgoldan) [reply treeherder/perf/models.py]: > > Regarding the global COUNTERS variable... I'm not sure that's the best approach.
> It could be read/updated not only by web requests, but by cronjobs also or even by one-off Dyno commands.
> 
> This is what the `COUNTERS_LOCK` is for:  It ensures that each thread in a process has exclusive accesses the COUNTERS.

Indeed, it ensures multiple threads from the **same** process don't step on each other's toes.
But I'm more concerned in multi process scenarios.

If we have 2 separate processes A & B, they'll both have their own COUNTERS object *(initially empty)*.
Even if they both start at the same time & lazy initialize COUNTERS from the database in the same way (such that both COUNTERS are initially the same), that doesn't guarantee the ids generated in A & B won't be duplicates; COUNTERS acts as a cache system at process level, built from scratch; A & B don't know each other's COUNTERS. When A & B will try to set these duplicated ids in the database, there are chances the client code will through errors.

In other words, A will think that "Hey! I can use the ids from this X chunk." B will consider that as well, for the very same X chunk. But when they try to sync that to the central database, one of them will fail.

**Comment by sarah-clements** [comment]
I'm not a good reviewer for this since it's out of my range of knowledge - @camd or @armenzg would be better.

One comment though, is that you said this will slow down all transactions. I think this could be a cause for concern since the log parser is already a time-consuming process. When you deployed this to prototype, were you able to see if there was a measurable difference in New Relic?

File: treeherder/config/settings.py [hunk 12]
```diff
@@ -159,6 +159,9 @@
         # truncated upon insertion, so Django highly recommends activating a strict mode for MySQL to
         # prevent data loss (either STRICT_TRANS_TABLES or STRICT_ALL_TABLES).
         'init_command': "SET sql_mode='STRICT_TRANS_TABLES'",
+        # Ensure we read the value inside the transaction, not some other volatile value
+        # https://docs.djangoproject.com/en/2.1/ref/databases/#mysql-isolation-level
```
Comment (armenzg) [inline treeherder/config/settings.py:163]: This link: https://docs.djangoproject.com/en/3.0/ref/databases/#mysql-isolation-level

File: treeherder/config/settings.py (hunk 11 above)
Comment (armenzg) [reply treeherder/config/settings.py:164]: I would wait for [bug 1603249](https://bugzilla.mozilla.org/show_bug.cgi?id=1603249) to land first.

How did the run in prototype go?
You can use New Relic to compare the data ingestion between that app and stage (those two apps have the same EC2 instance size and similar IOPS). You can put New Relic side by side for both apps and look at transactions by the slowest ones.

File: treeherder/perf/models.py [hunk 13]
```diff
@@ -256,6 +299,8 @@ class Meta:
         unique_together = ('repository', 'job', 'push', 'signature')
 
     def save(self, *args, **kwargs):
+        if not self.id:
+            self.id = self.next_id()
```
Comment (armenzg) [inline treeherder/perf/models.py:303]: Do I understand that this is to handle older rows?

File: treeherder/perf/models.py [hunk 14]
```diff
@@ -227,9 +229,50 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+    Used for tracking the blocks of ids severed to processes
+    Supports dense, globally unique, ids
+    """
+
+    model = models.CharField(max_length=80, primary_key=True)
+    chunk_size = models.IntegerField(null=False, default=100)
+    count = models.BigIntegerField(null=False)
+
+
+class AutoIncrementMixin:
```
Comment (armenzg) [inline treeherder/perf/models.py:243]: Would you mind adding a doc to this class explaining at a high level what it does? Thanks!

File: tests/model/test_cycle_data.py [hunk 15]
```diff
@@ -173,6 +173,7 @@ def test_cycle_job_with_performance_data(
     test_job.save()
 
     p = PerformanceDatum.objects.create(
+        id=1,
```
Comment (armenzg) [inline tests/model/test_cycle_data.py:176]: So all perf datums get created with a value of 1?

File: treeherder/perf/models.py (hunk 7 above)
Comment (klahnakoski) [reply treeherder/perf/models.py]: > When A & B will try to set these duplicated ids in the database, there are chances the client code will through errors.

You are correct.  That happens when the isolation level set to `REPEATABLE READ` (or less strict). When prototype was in that mode, occasionally two processes will have the correct timing, and both grab the same block at the same time.

With isolation level set to `SERIALIZABLE`, we get ACID compliance: We can assume only one process can enter the `with transaction.atomic()` section at a time:

* A requires a block of ids
* B requires a block of ids
* A reads the count (`N`) out of the database, and locks the record 
* B tries to reads the same count (`N`) out of the database, but is blocked 
* A increments by 100
* A updates the count in the database (`N+100), and releases lock
* B reads the count (`N+100`) out of the database
* B increments by 100
* B updates the count in the database (`N+200) 

Each process will have a unique block.

File: treeherder/config/settings.py (hunk 11 above)
Comment (klahnakoski) [reply treeherder/config/settings.py:164]: > I would wait for bug 1603249 to land first.

I am in no rush to land this

> How did the run in prototype go?

@armenzg  I did not know that prototype and stage could be compared.  I will deploy to prototype later this week and report my findings back here.  Thank you!

File: treeherder/perf/models.py [hunk 16]
```diff
@@ -227,9 +229,50 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
...
+                with transaction.atomic():
+                    db_counter = Counter.objects.filter(model=model_name).first()
+                    if not db_counter:
+                        # only happens once
+                        start = model.objects.aggregate(id=Max("id"))['id'] or 0
+                        db_counter = Counter.objects.create(
+                            model=model_name, chunk_size=100, count=start + 1
+                        )
+                    count = db_counter.count
+                    maxx = db_counter.count = count + db_counter.chunk_size
```
Comment (ionutgoldan) [inline treeherder/perf/models.py:267]: @armenzg how is Treeherder configured to handle requests, in terms of threads/processes?
Is it using thread/process pools? Is it spawning new processes per request, then recycles them?

If the latter case is true, then it's possible that each process grabs a chunk of size 100 from the database, uses less ids than that, then it terminates. The left ids will remain unused and unavailable, according to the `Counter` table.
@klahnakoski am I right about this? Are we expecting to waste some of the provided ids?

File: treeherder/perf/models.py (hunk 13 above)
Comment (klahnakoski) [reply treeherder/perf/models.py:303]: There are many instances of `PerformanceDatum.objects.create()`, none of which provide an `id`; they depend on the database `AUTO INCREMENT` to provide that.  This code replaces that feature: If there is no `id` then set it to a unique value.  For objects that already exist, the `id` already exists, and should not change.

**Comment by armenzg** [comment]
If possible please rebase over master so treeherder-admins does not get flagged for reviews. Thanks!

File: treeherder/perf/models.py (hunk 16 above)
Comment (klahnakoski) [reply treeherder/perf/models.py:282]: @ionutgoldan Yes, you got it correct. There will be holes in the id sequence when a process ends.  The maximum size of the holes is adjustable using the `chunk_size` which can be set dynamically in the database at any time.

**Comment by klahnakoski** [comment]
There must be a better way.  I posted a simplified version of my problem in Stackoverflow:

https://stackoverflow.com/questions/61523312/django3-using-mysql-read-and-update-in-single-transaction

File: treeherder/perf/models.py (hunk 7 above)
Comment (camd) [reply treeherder/perf/models.py]: I think an alternative would be the Django's builtin cache system, but I'm not at all familiar with it. @camd may be of help here.

It may be possible to use the cache to "lock" a set of ids.  But I would much prefer using the database to lock something.  That said, though, this is getting outside my domain knowledge.  :)  My db-fu only goes so deep and you guys are already deeper than that.  :)

**Comment by camd** [comment]
> There must be a better way. I posted a simplified version of my problem in Stackoverflow:
> 
> https://stackoverflow.com/questions/61523312/django3-using-mysql-read-and-update-in-single-transaction

I'll admit, that was my thinking, too.  :)  It just seems like this wheel must have already been invented.  But you, @klahnakoski, are **WAY** more db knowledgeable than I am.  :)  So I trust that what you come up with will be a good solution.  

I'm inclined to prefer the :wlach solution: https://bugzilla.mozilla.org/show_bug.cgi?id=1343328#c12

It feels safer to let the DB handle this.  But this is only my opinion.  If the migration is less than a day, then maybe a weekend planned downtime would be worth it.

**Comment by sarah-clements** [comment]
Something I just thought of.... there is ordering by `id` on the `PerformanceDatum table in the PerformanceDatumManager` and `ExtractPerf`.

File: treeherder/perf/models.py (hunk 16 above)
Comment (armenzg) [reply treeherder/perf/models.py:282]: I don't know Ionut. I know less about databases than people think.

**Comment by armenzg** [comment]
Would adding a new column with the new data type and backfill old data be good enough?
If we have 1M Perf datum rows we can start the new counter at 1M and fill that column for the older rows.

**Comment by sarah-clements** [comment]
> Would adding a new column with the new data type and backfill old data be good enough?
> If we have 1M Perf datum rows we can start the new counter at 1M and fill that column for the older rows.

Not sure I'm understanding this. Would older data would have two id numbers (their original id and the new id, in order to preserve the incremental ordering of id's)? I'm just wondering, in general, if the more complexity we add to this the higher risk of something going wrong versus switching to bigint and doing that migration (admittedly, I don't know what risks there are with that method).

**Comment by armenzg** [comment]
> > Would adding a new column with the new data type and backfill old data be good enough?
> > If we have 1M Perf datum rows we can start the new counter at 1M and fill that column for the older rows.
> 
> Not sure I'm understanding this. Would older data would have two id numbers (their original id and the new id, in order to preserve the incremental ordering of id's)?
> 
> I'm wondering, in general, if the more complexity we add to this the higher risk of something going wrong versus switching to bigint and doing that migration (admittedly, I don't know what risks there are with that method).

I think I agree with Sarah. I would rather use a TCW and get this done with.

**Comment by klahnakoski** [comment]
We will use a regular migration
